iOS: Fix multiple issues with PVRTC import, disable ETC1

DrMoriarty opened the pull request:
and some other VRAM compression formats.

For iOS we enable pvrtc feature by default for both GLES2/GLES3
Etc1 for iOS doesn't have any sense, so it disabled.
Fixed checks in export editor.
Fixed pvrtc ability detection in GLES2 driver.
Fixed pvrtc encoding procedure.
Compressed stream texture now saves right image dimensions for every compression format.
Also it allows to compressor resize image on it's side to make it more optimized.
So ETC1 format becames smaller then before. Pvrtc format always saves square texture.

*Bugsquad edit: closes #28683, closes #28621, closes #28596 and maybe others*
For #38040 and #35851, the fix has been moved to #42118.

DrMoriarty commented:
I'm not sure that it will work for master, so make PR for 3.2 branch.

akien-mga commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -5871,7 +5871,7 @@ void RasterizerStorageGLES2::initialize() {
 	config.s3tc_supported = config.extensions.has("GL_EXT_texture_compression_s3tc") || config.extensions.has("WEBGL_compressed_texture_s3tc");
 	config.etc1_supported = config.extensions.has("GL_OES_compressed_ETC1_RGB8_texture") || config.extensions.has("WEBGL_compressed_texture_etc1");
-	config.pvrtc_supported = config.extensions.has("IMG_texture_compression_pvrtc") || config.extensions.has("WEBGL_compressed_texture_pvrtc");
+	config.pvrtc_supported = config.extensions.has("GL_IMG_texture_compression_pvrtc") || config.extensions.has("IMG_texture_compression_pvrtc") || config.extensions.has("WEBGL_compressed_texture_pvrtc");
```
I think the one without `GL_` prefix might have been a bug, so it could likely be removed. All OpenGL extensions start with `GL_` IINM.

That's also what is used in the GLES3 backend:
```
config.pvrtc_supported = config.extensions.has("GL_IMG_texture_compression_pvrtc");
```
(BTW the `WEBGL_*` check should maybe also be added for GLES3?)

Filippopotamus commented:
Loading textures on iOS using GLES2 is still an issue. Is there a plan to merge this fix?

HEAVYPOLY commented:
+1 this is a big one for my project.

akien-mga commented:
I need contributors familiar with iOS to help review it, CC @bruvzg @naithar.

The fixes should also be ported to the `master` branch.

Filippopotamus commented:
@akien-mga I have cherry picked it to our 3.2.2 fork and it does seem to fix the GLES2 loading resource issue in iOS

akien-mga commented on editor/import/resource_importer_texture.cpp:
```diff
@@ -227,21 +227,8 @@ void ResourceImporterTexture::_save_stex(const Ref<Image> &p_image, const String
 	f->store_8('S');
 	f->store_8('T'); //godot streamable texture
 
-	bool resize_to_po2 = false;
-
-	if (p_compress_mode == COMPRESS_VIDEO_RAM && p_force_po2_for_compressed && (p_mipmaps || p_texture_flags & Texture::FLAG_REPEAT)) {
-		resize_to_po2 = true;
-		f->store_16(next_power_of_2(p_image->get_width()));
-		f->store_16(p_image->get_width());
-		f->store_16(next_power_of_2(p_image->get_height()));
-		f->store_16(p_image->get_height());
-	} else {
-		f->store_16(p_image->get_width());
-		f->store_16(0);
-		f->store_16(p_image->get_height());
-		f->store_16(0);
-	}
-	f->store_32(p_texture_flags);
```
These are the only changes I'm not confident with as they impact all formats, so I want @reduz to review this more closely.

The rest of the changes seem good though (with some obvious bugfixes like in `drivers/gles2` or `platform/iphone/export/export.cpp`), so I'd suggest splitting the texture importer change to a dedicated PR and remove it here, so that the rest can be merged first.

DrMoriarty committed:
iOS: Fix multiple issues with PVRTC import, disable ETC1

Fixes: #28683, #28621, #28596 and maybe others

For iOS we enable pvrtc feature by default for both GLES2/GLES3
Etc1 for iOS doesn't have any sense, so it disabled.
Fixed checks in export editor.
Fixed pvrtc ability detection in GLES2 driver.
Fixed pvrtc encoding procedure.

akien-mga commented:
Rebased to remove the unrelated CHANGELOG conflict, and made the commit log a bit more explicit.

Could you check if the issues referenced as fixed by this PR are still all fixed, or if some of them are fixed by #42118 instead (or only fixed once both are merged)?

akien-mga commented on editor/editor_export.cpp:
```diff
@@ -1298,6 +1298,31 @@ String EditorExportPlatform::test_etc2() const {
+
+	String driver = ProjectSettings::get_singleton()->get("rendering/quality/driver/driver_name");
+	bool driver_fallback = ProjectSettings::get_singleton()->get("rendering/quality/driver/fallback_to_gles2");
+	bool etc_supported = ProjectSettings::get_singleton()->get("rendering/vram_compression/import_etc");
```
I guess `etc_supported` is not relevant for iOS (and now disabled by `platform/iphone/export/export.cpp` changes), so maybe it can be removed from the checks / error messages?

DrMoriarty replied:
Yes, I think so

akien-mga replied:
I pushed an amend that removes the `etc_supported` check.

DrMoriarty commented:
@akien-mga 
issues https://github.com/godotengine/godot/issues/38040 and https://github.com/godotengine/godot/issues/35851 related to PR #42118 
Other related to this PR

akien-mga reviewed (approved):
Looks great to me. I can't really assess the `modules/pvr/texture_loader_pvr.cpp` changes but I trust that you tested them and that they make things work as they should. The rest is :+1:

akien-mga commented:
I forward-ported the relevant changes for the `master` branch with #42262 (untested, and some of it might not make sense for Vulkan, hence the TODOs).

reduz commented:
As most users dont go iOS out of the box, I think disabling PVRTC by default is likely the best idea, and only enable it if you are planning to export to iOS.

DrMoriarty commented:
@reduz this PR don't change default enabled compressed textures formats. So PVRTC will be disabled until user enabled it.

akien-mga commented:
Yes @reduz seems to have misinterpreted the code, this doesn't change the fact that PVRTC is off by default (but it should be toggled on if you want to target iOS with GLES2, just like ETC1 needs to be toggled on manually if you want to target Android with GLES2).

akien-mga commented:
Thanks!
